### PR TITLE
#12 & #21 -Backend and Frontend view tenders - Aisha Athman Lali

### PR DESCRIPTION
**Pull request by aishaathmanlali**
## Description

This is a PR for viewing tenders in both backend and frontend.

## Related to

#12 and #21

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have carefully reviewed my own code
- [ ] I have commented my code
- [ ] I have updated any documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,7 +12,11 @@ Ensure the following are installed on your system before proceeding:
 
 This is our deployed website link https://love-me-tender-51qa.onrender.com
 
-The render is on auto-deploy so it should render new changes to the website.
+To get the latest version deployed on render;
+
+1. Go to render dashboard
+2. Click on manual deploy button
+3. Select deploy latest commit
```
Comment (soonick) [inline README.md]: Thanks for updating this. In the future, please add it in a different commit, since it's unrelated to the other changes

File: client/src/TenderList.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+	const fetchTenders = async (pageNumber) => {
+		setLoading(true);
+		try {
+			const response = await fetch(`/api/tenders?page=${pageNumber}`);
+			const data = await response.json();
+			setTenders((prevTenders) => [...prevTenders, ...data.tenders]);
+			setHasMore(data.tenders.length === 25);
+		} catch (error) {
+			console.error("Error fetching tenders:", error);
```
Comment (soonick) [inline client/src/TenderList.js]: Don't log errors. Inform the user.

File: client/src/TenderList.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	const [page, setPage] = useState(1);
+	const [hasMore, setHasMore] = useState(true);
+
+	const fetchTenders = async (pageNumber) => {
+		setLoading(true);
+		try {
+			const response = await fetch(`/api/tenders?page=${pageNumber}`);
+			const data = await response.json();
+			setTenders((prevTenders) => [...prevTenders, ...data.tenders]);
+			setHasMore(data.tenders.length === 25);
```
Comment (soonick) [inline client/src/TenderList.js]: If the page has 25 items, then this doesn't necessarily mean there are more items. It's better if the server explicitly tells the client if there are more items

File: client/src/TenderList.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+									tender["Tender Project Deadline Date"]
+								).toLocaleDateString()}
+							</td>
+							<td>{tender["Tender Status"]}</td>
+						</tr>
+					))}
+				</tbody>
+			</table>
+			{loading && <p>Loading...</p>}
+			{hasMore && !loading && <button onClick={loadMore}>Load More</button>}
```
Comment (soonick) [inline client/src/TenderList.js]: Load more, makes it sound like more items are going to be added inline. I think `next page` is clearer

File: client/src/ViewTendersButton.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import React from "react";
+
+const ViewTendersButton = ({ onClick }) => (
```
Comment (soonick) [inline client/src/ViewTendersButton.js]: Why did you create a custom component for this? It seems like it doesn't add anything to a plan button

File: client/src/pages/BuyerDashboard.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { Link } from "react-router-dom";
+// import "./BuyerDashboard.css";
```
Comment (soonick) [inline client/src/pages/BuyerDashboard.js]: Remove

File: database/database.json [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
-    }
+	"dev": {
+		"driver": "pg",
+		"user": "postgres",
+		"password": "opensesame",
+		"host": "localhost",
+		"database": "cyf_tms",
+		"port": "5432"
+	}
 }
```
Comment (soonick) [inline database/database.json]: Seems like you didn't change anything here. Remove this file from commit

File: database/migrations/20240715163245-initial-db.js [hunk 8]
```diff
@@ -1,51 +1,61 @@
-var dbm;
-var type;
-var seed;
-var fs = require('fs');
-var path = require('path');
-var Promise;
+// let dbm;
```
Comment (soonick) [inline database/migrations/20240715163245-initial-db.js]: This file is auto-generated. You shouldn't need to modify it

File: docker-compose.yml [hunk 9]
```diff
@@ -14,3 +14,9 @@ services:
       POSTGRES_PASSWORD: "opensesame"
     ports:
       - "5432:5432"
+    networks:
+      - default
+
+networks:
+  default:
+    name: workspace_default
```
Comment (soonick) [inline docker-compose.yml]: Is this necessary? I think I was able to get it to work without this

File: server/api.js [hunk 10]
```diff
@@ -1,9 +1,10 @@
 import { Router } from "express";
+import { query } from "./db";
 
 const router = Router();
 
 router.get("/", (_, res) => {
-	res.status(200).json({ message: "WELCOME TO LOVE ME TENDER SITE" });
+	res.status(200).json({ message: "TENDER MANAGEMENT SYSTEM" });
```
Comment (soonick) [inline server/api.js]: Remove this change. It's unrelated to the commit and probably unnecessary

File: server/api.js [hunk 11]
```diff
@@ -134,4 +135,33 @@ router.post("/publish-tenders", (req, res) => {
...
+		const result = await query(
+			`
+		SELECT
+			t.id AS "Tender ID",
+			t.title AS "Tender Title",
+			t.creation_date AS "Tender Created Date",
+			t.announcement_date AS "Tender Announcement Date",
+			t.deadline AS "Tender Closing Date",
+			t.deadline AS "Tender Project Deadline Date",
+			t.status AS "Tender Status"
```
Comment (soonick) [inline server/api.js]: I don't think it's a good idea to create these aliases. Let the client decide the headers to user for the table

File: server/api.js [hunk 12]
```diff
@@ -134,4 +135,33 @@ router.post("/publish-tenders", (req, res) => {
...
+		ORDER BY
+			t.creation_date DESC
+		LIMIT $1 OFFSET $2;
+		`,
+			[limit, offset]
+		);
+
+		res.status(200).json({ tenders: result.rows });
+	} catch (err) {
+		res.status(500).json({ error: err.message });
```
Comment (soonick) [inline server/api.js]: Don't send the raw error message to the client. It might contain sensitive information. Just send a generic server error code.

File: server/db.js [hunk 13]
```diff
@@ -1,12 +1,13 @@
-import { Pool } from "pg";
-
+import pg from "pg";
```
Comment (soonick) [inline server/db.js]: Are the changes in this file necessary? Seems like nothing really changed

File: server/mail.js [hunk 14]
```diff
@@ -2,23 +2,23 @@ import nodemailer from "nodemailer";
...
-        html: message,
-    });
+	return await transporter.sendMail({
+		from: "no_reply@example.com",
+		to: recipient,
+		subject,
+		text: message,
+		html: message,
+	});
 };
```
Comment (soonick) [inline server/mail.js]: Seems like nothing really changed. Remove from commit

**Review by soonick — CHANGES_REQUESTED** [review]
In the future, separate your backend commit from your frontend commit. It makes it easier to review

File: README.md (hunk 1 above)
Comment (aishaathmanlali) [reply README.md]: Okay next time I will make a different commit iof its different from the task

Comment (soonick) [reply README.md]: It's probably good practice to do it now. Create a new commit just with this change and remove it from this commit.

File: server/mail.js (hunk 14 above)
Comment (aishaathmanlali) [reply server/mail.js]: done

File: server/db.js (hunk 13 above)
Comment (aishaathmanlali) [reply server/db.js]: I have reverted the change

File: server/api.js (hunk 12 above)
Comment (aishaathmanlali) [reply server/api.js]: I have updated it

File: server/api.js (hunk 11 above)
Comment (aishaathmanlali) [reply server/api.js]: I have modified it

File: server/api.js (hunk 10 above)
Comment (aishaathmanlali) [reply server/api.js]: I have changed it

File: client/src/ViewTendersButton.js (hunk 5 above)
Comment (aishaathmanlali) [reply client/src/ViewTendersButton.js]: I have removed and added a plain button

File: client/src/TenderList.js (hunk 4 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: I have changed it to Next Page

File: client/src/TenderList.js (hunk 2 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: I have changed it

**Comment by aishaathmanlali** [comment]
> In the future, separate your backend commit from your frontend commit. It makes it easier to review

I will do that. Thank you

File: client/src/TenderList.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+		setLoading(true);
+		try {
+			const response = await fetch(`/api/tenders?page=${pageNumber}`);
+			if (!response.ok) {
+				throw new Error(`HTTP error! status: ${response.status}`);
+			}
+			const data = await response.json();
+
+			console.log("Fetched data:", data);
+
```
Comment (soonick) [inline client/src/TenderList.js]: Remove

File: client/src/TenderList.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+			if (Array.isArray(data.tenders)) {
+				setTenders((prevTenders) => [...prevTenders, ...data.tenders]);
+				setHasMore(data.tenders.length === 25);
+				setError(null);
+			} else {
+				throw new Error("Invalid data format: data.tenders is not an array");
```
Comment (soonick) [inline client/src/TenderList.js]: This message is irrelevant and confusing to a user. Just tell them there was a server error

File: client/src/TenderList.js [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+			const response = await fetch(`/api/tenders?page=${pageNumber}`);
+			if (!response.ok) {
+				throw new Error(`HTTP error! status: ${response.status}`);
+			}
+			const data = await response.json();
+
+			console.log("Fetched data:", data);
+
+			if (Array.isArray(data.tenders)) {
+				setTenders((prevTenders) => [...prevTenders, ...data.tenders]);
```
Comment (soonick) [inline client/src/TenderList.js]: Are we just appending tenders in the same page?

File: client/src/TenderList.js [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+				setHasMore(data.tenders.length === 25);
```
Comment (soonick) [inline client/src/TenderList.js]: Same issue as before. This doesn't mean there are more tenders.

File: client/src/TenderList.js [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+						<th>Tender Title</th>
+						<th>Tender Created Date</th>
+						<th>Tender Announcement Date</th>
+						<th>Tender Closing Date</th>
+						<th>Tender Project Deadline Date</th>
+						<th>Tender Status</th>
+					</tr>
+				</thead>
+				<tbody>
+					{console.log(tenders)}
```
Comment (soonick) [inline client/src/TenderList.js]: Remove all `console.log`s from your code

File: client/src/ViewTendersButton.js (hunk 5 above)
Comment (soonick) [reply client/src/ViewTendersButton.js]: You need to delete this file

**Review by soonick — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (aishaathmanlali) [reply README.md]: I have committed it separately

File: client/src/ViewTendersButton.js (hunk 5 above)
Comment (aishaathmanlali) [reply client/src/ViewTendersButton.js]: done

File: client/src/TenderList.js (hunk 15 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: done

File: client/src/TenderList.js (hunk 16 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: I have changed it

File: client/src/TenderList.js (hunk 17 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: after 25 items it should display the second page however I have not succeeded in doing that yet

File: client/src/TenderList.js (hunk 18 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: I have added backend for when there are more than 25 items to be displayed

File: client/src/TenderList.js (hunk 19 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: I have done that

File: database/database.json (hunk 7 above)
Comment (aishaathmanlali) [reply database/database.json]: done

File: docker-compose.yml (hunk 9 above)
Comment (aishaathmanlali) [reply docker-compose.yml]: removed

File: client/src/TenderList.js (hunk 3 above)
Comment (aishaathmanlali) [reply client/src/TenderList.js]: I have done that

File: client/src/TenderList.js [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+	try {
+		const response = await fetch(`/api/tenders?page=${page}`);
+		if (!response.ok) {
+			throw new Error(`HTTP error! status: ${response.status}`);
+		}
+		const data = await response.json();
+
+		if (Array.isArray(data.resource.tenders)) {
+			setTenders(data.resource.tenders);
+			setPagination(data.resource.pagination);
```
Comment (soonick) [inline client/src/TenderList.js]: `pagination` should be directly in the data, not inside the `resource`

File: client/src/TenderList.js [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+	});
+	const [error, setError] = useState(null);
+	const navigate = useNavigate();
+
+	const fetchTenders = async (page) => {
+		setLoading(true);
+	try {
+		const response = await fetch(`/api/tenders?page=${page}`);
+		if (!response.ok) {
+			throw new Error(`HTTP error! status: ${response.status}`);
```
Comment (soonick) [inline client/src/TenderList.js]: The user doesn't need to know the status code. Imagine how a user would react if they went to a page, and they saw `Status code: 500`. It doesn't tell them anything.

File: client/src/TenderList.js [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+							<td>{tender.status}</td>
+						</tr>
+					))}
+				</tbody>
+			</table>
+			{loading && <p>Loading...</p>}
+			<div>
+				<button
+					onClick={loadPreviousPage}
+					disabled={loading || pagination.currentPage <= 1}
```
Comment (soonick) [inline client/src/TenderList.js]: If there are no previous pages, it should not be shown

File: client/src/TenderList.js [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+				>
+					Previous Page
+				</button>
+				<button
+					onClick={loadNextPage}
+					disabled={loading || pagination.currentPage >= pagination.totalPages}
```
Comment (soonick) [inline client/src/TenderList.js]: If there are no more pages, it should not be shown

File: server/api.js [hunk 24]
```diff
@@ -71,70 +71,6 @@ router.post("/publish-tenders", (req, res) => {
 	res.status(200).json({ message: "Form submitted successfully!" });
 });
 
-router.get("/skills", (req, res) => {
```
Comment (soonick) [inline server/api.js:74]: Why are you removing all of this?

File: server/api.js [hunk 25]
```diff
@@ -164,4 +100,39 @@ router.get("/bidder-bid", async (req, res) => {
 			.send({ code: "SERVER_ERROR" });
 });
 
+router.get("/tenders", async (req, res) => {
+	const page = parseInt(req.query.page, 10) || 1;
+	const limit = parseInt(req.query.limit, 10) || 25;
```
Comment (soonick) [inline server/api.js]: Remove this. The client is not sending it. We can hard code it to 25 for now.

File: server/api.js [hunk 26]
```diff
@@ -164,4 +100,39 @@ router.get("/bidder-bid", async (req, res) => {
...
+	`;
+	try {
+		const countResult = await db.query(countSql);
+		const totalItems = parseInt(countResult.rows[0].count, 10);
+		const totalPages = Math.ceil(totalItems / limit);
+
+		const dataResult = await db.query(dataSql, [limit, offset]);
+		const tenders = dataResult.rows;
+
+	res.status(200).json({
```
Comment (soonick) [inline server/api.js]: Indentation is not correct. Is your precommit hook working? It shouldn't have let you commit this.

File: server/api.js [hunk 27]
```diff
@@ -164,4 +100,39 @@ router.get("/bidder-bid", async (req, res) => {
...
+		resource: {
+			tenders,
+			pagination: {
+				itemsPerPage: limit,
+				currentPage: page,
+				totalPages,
+			},
+		},
```
Comment (soonick) [inline server/api.js]: This doesn't follow the standard I put in #launch-module-web-wizards. Please verify it

File: server/db.js [hunk 28]
```diff
@@ -32,8 +31,8 @@ export const disconnectDb = () => pool.end();
  * `await db.query("<SQL>", [...<variables>])`.
  */
 export default {
-	query: (...args) => {
-		logger.debug("Postgres querying %O", args);
-		return pool.query.apply(pool, args);
+	query: (text, params) => {
+		logger.debug("Postgres querying %O", { text, params });
+		return pool.query(text, params);
```
Comment (soonick) [inline server/db.js]: I don't think this is the same thing. Why did you change it? Are you sure this doesn't break anything?

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 24 above)
Comment (aishaathmanlali) [reply server/api.js:74]: it was repeating

File: server/api.js (hunk 26 above)
Comment (aishaathmanlali) [reply server/api.js]: yes it is working

File: server/db.js (hunk 28 above)
Comment (aishaathmanlali) [reply server/db.js]: I have changed it back to the way it was

File: README.md [hunk 29]
```diff
@@ -12,7 +12,11 @@ Ensure the following are installed on your system before proceeding:
 
 This is our deployed website link https://love-me-tender-51qa.onrender.com
 
-The render is on auto-deploy so it should render new changes to the website.
+To upload the project's latest version follow the steps below;
```
Comment (soonick) [inline README.md]: nit. `To deploy the ...`

File: README.md [hunk 30]
```diff
@@ -38,23 +42,21 @@ npm install
 
 4. To run the migrations, navigate to database folder and run
 
-```sh
+````sh
```
Comment (soonick) [inline README.md]: This is broken

File: README.md [hunk 31]
```diff
@@ -38,23 +42,21 @@ npm install
...
 node ../node_modules/db-migrate/bin/db-migrate up
-```
 
 To modify the database, run the command below from the database folder, and add DB modifications in the generated ...-up.sql file
 
 Then run above command again.
 
 ```sh
 node ../node_modules/db-migrate/bin/db-migrate create name-for-the-migration --sql-file
-```
```
Comment (soonick) [inline README.md]: This is broken

File: README.md [hunk 32]
```diff
@@ -38,23 +42,21 @@ npm install
...
 
 5. Run the build command
 
 ```sh
 npm run build
-```
+````
```
Comment (soonick) [inline README.md]: This is broken

**Review by soonick — CHANGES_REQUESTED** [review]
Looks good. Just need to fix some small things that broke in the README

**Review by soonick — APPROVED** [review]
